streams/dshr_strdata_mod.F90

```diff
@@ -1225,123 +1422,230 @@ subroutine shr_strdata_readstrm(sdat, ns, stream, stream_mesh, &
+          write(sdat%logunit,F02)' reading '//&
+               trim(per_stream%fldlist_stream(nf))//' into '//trim(per_stream%fldlist_model(nf)),&
                ' at time index: ',nt
        end if
 
+       ! read the data
        call pio_setframe(pioid, varid, int(nt,kind=Pio_Offset_Kind))
 
        if (pio_iovartype == PIO_REAL) then
-          if (pio_iodesc_set) then
-             call pio_read_darray(pioid, varid, pio_iodesc, data_real, rcode)
-          else
-             rcode = pio_get_var(pioid, varid,start=(/1,1,nt/),count=(/1,1,1/), ival=data_real)
-          endif
-          if ( rcode /= PIO_NOERR ) then
-             call shr_sys_abort(' ERROR: reading in variable: '// trim(fldlist_stream(nf)))
+          ! -----------------------------
+          ! pio_iovartype is PIO_REAL
+          ! -----------------------------
+          if (stream_nlev > 1) then
+             if (per_stream%stream_pio_iodesc_set) then
+                call pio_read_darray(pioid, varid, per_stream%stream_pio_iodesc, data_real2d, rcode)
+             else
+                rcode = pio_get_var(pioid, varid,start=(/1,1,1,nt/),count=(/1,1,1,1/), ival=data_real2d)
+             end if
+             if ( rcode /= PIO_NOERR ) then
+                call shr_sys_abort(' ERROR: reading in variable: '// trim(per_stream%fldlist_stream(nf)))
+             end if
+             if (handlefill) then
+                do l = 1,stream_nlev
+                   do n = 1,size(dataptr2d, dim=2)
+                      if (.not. shr_infnan_isnan(data_real2d(n,l)) .and. data_real2d(n,l) .ne. fillvalue_r4) then
+                         dataptr2d(l,n) = real(data_real2d(n,l), kind=r8) ! Note the order of indices
+                      else
+                         dataptr2d(l,n) = r8fill
+                      endif
+                   enddo
+                end do
+             else
+                do l = 1,stream_nlev
+                   do n = 1,size(dataptr2d, dim=2)
+                      dataptr2d(l,n) = real(data_real2d(n,l), kind=r8)
+                   end do
+                end do
+             end if
+          else ! stream_nlev == 1
+             if (per_stream%stream_pio_iodesc_set) then
+                call pio_read_darray(pioid, varid, per_stream%stream_pio_iodesc, data_real1d, rcode)
+             else
+                rcode = pio_get_var(pioid, varid,start=(/1,1,nt/),count=(/1,1,1/), ival=data_real1d)
+             endif
+             if ( rcode /= PIO_NOERR ) then
+                call shr_sys_abort(' ERROR: reading in variable: '// trim(per_stream%fldlist_stream(nf)))
+             end if
+             if (handlefill) then
+                do n=1,size(dataptr1d)
+                   if(.not. shr_infnan_isnan(data_real1d(n)) .and. data_real1d(n) .ne. fillvalue_r4) then
+                      dataptr1d(n) = real(data_real1d(n), kind=r8)
+                   else
+                      dataptr1d(n) = r8fill
```

Comment:
Why are there both fillvalue_r8 and r8fill?

Reply:
You put this logic in. I was only extending it. I think the reason is that r8fill is shr_const_spval and fillvalue_r8 is read in the from the file.